### DISCUSSION TITLE
tquic在回调函数外创建流的问题

### DISCUSSION THREAD

**Discussion by LapseLove** [Tencent/tquic#431] [Q&A (问题和回答)] [answered]
我在c示例simple_client中另外添加了一个ev_io_watcher监听stdin输入并且把其设置成非阻塞，想做到在stdin有数据时就接收然后使用quic_stream_uni_new创建一个新的流把数据发出去。
但是输入数据并且确实执行了quic_stream_uni_new后on_stream_created并没有触发，而且把创建的流id打印出来以后值非常大为124176020812530。
而上一个在on_connection_established回调函数里面创建的流就可以正常触发on_stream_created发送数据。
请问这是什么问题呢？我不可以在回调函数外使用quic_stream_uni_new吗？

**Comment by iyangsj** [comment, marked as answer]
> 把创建的流id打印出来以后值非常大

检查下quic_stream_uni_new()返回值，是否返回了错误


> 我不可以在回调函数外使用quic_stream_uni_new吗

可以，但需要额外手工调用 `quic_endpoint_process_connections()`

简单的做法是仅在回调函数中执行这类函数

**Reply by LapseLove** [reply]
手工调用quic_endpoint_process_connections()解决了流创建的问题，但是当服务端创建单向流返回数据给客户端接收后，客户端就又开始无法创建流了，quic_stream_uni_new()的返回值为-5，查看RFC 9000上面的说明是Frame received in invalid stream state。但是这个时候客户端已经接收完了服务端的数据，应该是没有继续收到数据了，请问这个是什么问题呢？

**Reply by iyangsj** [reply]
返回值-5 代表 StreamStateError错误
https://github.com/Tencent/tquic/blob/develop/src/error.rs#L192

可以把你的示例程序贴出来看看（可以上传附件）

**Reply by LapseLove** [reply]
[client_github.txt](https://github.com/user-attachments/files/17855191/client_github.txt)
[server_github.txt](https://github.com/user-attachments/files/17855194/server_github.txt)
上面是示例程序的代码，下面是输入后发生错误的截图
![服务器](https://github.com/user-attachments/assets/f5a8bd52-3bc4-4855-b15c-2c57e1171a57)
![客户端](https://github.com/user-attachments/assets/af7cb248-5b12-4740-b2d0-d94be0cf1626)
